src/jmespath/Parser.js

```diff
@@ -150,6 +150,14 @@ export default class Parser {
     this.index += 1;
   }
 
+  _getIndex() {
```

Comment:
A level of backtracking was required to keep the code abstracted. This was needed because to distinguish between `[expression1: expression2]` and `[expression1, expression2]` we need to first parse the first expression.

It was cleaner to go back to the original position after determining the type instead of proceeding ahead from the current position.